File: udon_decompiler/analysis/ir/control_flow_graph.py
from __future__ import annotations

from typing import Dict, Optional, Set

from udon_decompiler.analysis.ir.control_flow_node import ControlFlowNode
from udon_decompiler.analysis.ir.dominance import (
    compute_dominance,
    mark_nodes_with_reachable_exits,
)
from udon_decompiler.analysis.ir.nodes import (
    IRBlock,
    IRBlockContainer,
    IRHighLevelDoWhile,
    IRHighLevelSwitch,
    IRHighLevelWhile,
    IRIf,
    IRJump,
    IRLeave,
    IRReturn,
    IRStatement,
    IRSwitch,
)


class ControlFlowGraph:
    """
    ILSpy-style control-flow graph for one IRBlockContainer.

    A CFG snapshot is built once before block transforms for the container.
    Later transforms may move/reorder blocks, but this graph keeps the original
    node identity/mapping for algorithms that rely on a stable snapshot.
    """

    def __init__(
        self,
        container: IRBlockContainer,
        function_body: IRBlockContainer,
    ) -> None:
        self.container = container
        self._function_body = function_body

        self.cfg: list[ControlFlowNode] = []
        self._dict: Dict[IRBlock, ControlFlowNode] = {}

        self._node_has_direct_exit_out_of_container: list[bool] = []
        self._node_has_reachable_exit: list[bool] = []

        self._container_block_to_index: Dict[IRBlock, int] = {
            block: index for index, block in enumerate(container.blocks)
        }
        self._nested_descendant_blocks = self._collect_nested_descendant_blocks(
            container
        )

        self._initialize_nodes()
        self._create_edges()

        if self.cfg:
            entry = self.cfg[0]
            compute_dominance(entry)
            self._node_has_reachable_exit = mark_nodes_with_reachable_exits(self.cfg)
            leaving = self._find_nodes_with_exits_out_of_container()
            self._node_has_reachable_exit = [
                self._node_has_reachable_exit[i] or leaving[i]
                for i in range(len(self.cfg))
            ]

    def get_node(self, block: IRBlock) -> ControlFlowNode:
        return self._dict[block]

    def has_reachable_exit(self, node: ControlFlowNode) -> bool:
        self._validate_node(node)
        return self._node_has_reachable_exit[node.user_index]

    def has_direct_exit_out_of_container(self, node: ControlFlowNode) -> bool:
        self._validate_node(node)
        return self._node_has_direct_exit_out_of_container[node.user_index]

    def _validate_node(self, node: ControlFlowNode) -> None:
        if node.user_index < 0 or node.user_index >= len(self.cfg):
            raise IndexError("ControlFlowNode index out of range")
        if self.cfg[node.user_index] is not node:
            raise ValueError("ControlFlowNode does not belong to this CFG")

    def _initialize_nodes(self) -> None:
        for index, block in enumerate(self.container.blocks):
            node = ControlFlowNode(user_index=index, block=block)
            self.cfg.append(node)
            self._dict[block] = node

        self._node_has_direct_exit_out_of_container = [False] * len(self.cfg)
        self._node_has_reachable_exit = [False] * len(self.cfg)

    def _create_edges(self) -> None:
        for index, block in enumerate(self.container.blocks):
            source_node = self.cfg[index]
            descendant_container_ids = self._collect_descendant_container_ids(block)
            for statement in block.statements:
                self._process_statement(
                    source_node=source_node,
                    source_block=block,
                    statement=statement,
                    descendant_container_ids=descendant_container_ids,
                )
            self._add_fallthrough_edge_if_needed(
                source_node=source_node,
                block=block,
                block_index=index,
            )

    def _add_fallthrough_edge_if_needed(
        self,
        source_node: ControlFlowNode,
        block: IRBlock,
        block_index: int,
    ) -> None:
        if self._block_has_unreachable_endpoint(block):
            return

        next_index = block_index + 1
        if next_index < len(self.cfg):
            source_node.add_edge_to(self.cfg[next_index])
            return

        # Falling off the end of the current container exits the region.
        self._node_has_direct_exit_out_of_container[source_node.user_index] = True

    def _process_statement(
        self,
        source_node: ControlFlowNode,
        source_block: IRBlock,
        statement: IRStatement,
        descendant_container_ids: Set[int],
    ) -> None:
        if isinstance(statement, IRJump):
            self._process_branch_target(source_node, statement.target)
            return

        if isinstance(statement, IRIf):
            self._process_statement(
                source_node,
                source_block,
                statement.true_statement,
                descendant_container_ids,
            )
            if statement.false_statement is not None:
                self._process_statement(
                    source_node,
                    source_block,
                    statement.false_statement,
                    descendant_container_ids,
                )
            return

        if isinstance(statement, IRSwitch):
            for target in statement.cases.values():
                self._process_branch_target(source_node, target)
            if statement.default_target is not None:
                self._process_branch_target(source_node, statement.default_target)
            return

        if isinstance(statement, IRLeave):
            if self._is_leave_out_of_container(
                source_block=source_block,
                leave=statement,
                descendant_container_ids=descendant_container_ids,
            ):
                self._node_has_direct_exit_out_of_container[source_node.user_index] = (
                    True
                )
            return

        if isinstance(statement, IRBlockContainer):
            for nested_block in statement.blocks:
                for nested_statement in nested_block.statements:
                    self._process_statement(
                        source_node,
                        source_block,
                        nested_statement,
                        descendant_container_ids,
                    )

    def _block_has_unreachable_endpoint(self, block: IRBlock) -> bool:
        if not block.statements:
            return False
        return self._statement_has_unreachable_endpoint(block.statements[-1])

    def _statement_has_unreachable_endpoint(self, statement: IRStatement) -> bool:
        if isinstance(statement, (IRJump, IRLeave, IRReturn, IRSwitch)):
            return True

        if isinstance(statement, IRIf):
            if statement.false_statement is None:
                return False
            return (
                self._statement_has_unreachable_endpoint(statement.true_statement)
                and self._statement_has_unreachable_endpoint(
                    statement.false_statement
                )
            )

        if isinstance(statement, IRBlock):
            if not statement.statements:
                return False
            return self._statement_has_unreachable_endpoint(statement.statements[-1])

        if isinstance(statement, IRHighLevelSwitch):
            return True

        if isinstance(statement, (IRHighLevelWhile, IRHighLevelDoWhile)):
            return False

        return False

    def _process_branch_target(
        self,
        source_node: ControlFlowNode,
        target: IRBlock,
    ) -> None:
        if target in self._container_block_to_index:
            target_index = self._container_block_to_index[target]
            source_node.add_edge_to(self.cfg[target_index])
            return

        # Internal control flow within nested containers is ignored for this CFG.
        if target in self._nested_descendant_blocks:
            return

        self._node_has_direct_exit_out_of_container[source_node.user_index] = True

    def _is_leave_out_of_container(
        self,
        source_block: IRBlock,
        leave: IRLeave,
        descendant_container_ids: Set[int],
    ) -> bool:
        # Returning from function is not considered a reachable exit.
        if leave.target_container is self._function_body:
            return False

        # Leave into containers nested inside this block is internal control flow.
        if id(leave.target_container) in descendant_container_ids:
            return False

        # Everything else exits the current container region.
        _ = source_block
        return True

    def _find_nodes_with_exits_out_of_container(self) -> list[bool]:
        leaving = [False] * len(self.cfg)
        for node in self.cfg:
            if leaving[node.user_index]:
                continue
            if not self._node_has_direct_exit_out_of_container[node.user_index]:
                continue

            current: Optional[ControlFlowNode] = node
            while current is not None:
                if leaving[current.user_index]:
                    break
                leaving[current.user_index] = True
                current = current.immediate_dominator

        return leaving

    def _collect_nested_descendant_blocks(
        self,
        container: IRBlockContainer,
    ) -> Set[IRBlock]:
        descendants: set[IRBlock] = set()
        visited_containers: set[int] = set()

        def visit_statement(statement: IRStatement) -> None:
            if not isinstance(statement, IRBlockContainer):
                return
            visit_container(statement)

        def visit_container(nested: IRBlockContainer) -> None:
            nested_id = id(nested)
            if nested_id in visited_containers:
                return
            visited_containers.add(nested_id)

            for nested_block in nested.blocks:
                descendants.add(nested_block)
                for nested_statement in nested_block.statements:
                    visit_statement(nested_statement)

        for block in container.blocks:
            for statement in block.statements:
                visit_statement(statement)

        return descendants

    def _collect_descendant_container_ids(self, block: IRBlock) -> Set[int]:
        result: set[int] = set()
        visited_containers: set[int] = set()

        def visit_statement(statement: IRStatement) -> None:
            if not isinstance(statement, IRBlockContainer):
                return

            nested = statement
            nested_id = id(nested)
            if nested_id in visited_containers:
                return
            visited_containers.add(nested_id)

            result.add(nested_id)
            for nested_block in nested.blocks:
                for nested_statement in nested_block.statements:
                    visit_statement(nested_statement)

        for statement in block.statements:
            visit_statement(statement)

        return result

File: src/decompiler/ir/control_flow_graph.rs
use std::collections::{HashMap, HashSet};

use super::control_flow_node::ControlFlowNode;
use super::dominance::{compute_dominance, mark_nodes_with_reachable_exits};
use super::nodes::{
    IrBlock, IrBlockContainer, IrHighLevelDoWhile, IrHighLevelSwitch, IrHighLevelWhile, IrIf,
    IrJump, IrLeave, IrReturn, IrStatement, IrSwitch,
};

#[derive(Debug, Clone)]
pub struct ControlFlowGraph {
    pub container_id: u32,
    pub function_body_container_id: u32,
    pub nodes: Vec<ControlFlowNode>,

    block_index_by_address: HashMap<u32, usize>,
    node_has_direct_exit_out_of_container: Vec<bool>,
    node_has_reachable_exit: Vec<bool>,
}

impl ControlFlowGraph {
    pub fn new(container: &IrBlockContainer, function_body_container_id: u32) -> Self {
        let mut out = Self {
            container_id: container.id,
            function_body_container_id,
            nodes: Vec::new(),
            block_index_by_address: HashMap::new(),
            node_has_direct_exit_out_of_container: Vec::new(),
            node_has_reachable_exit: Vec::new(),
        };

        out.initialize_nodes(container);
        out.create_edges(container);

        if !out.nodes.is_empty() {
            compute_dominance(0, &mut out.nodes);
            out.node_has_reachable_exit = mark_nodes_with_reachable_exits(&out.nodes);
            let leaving = out.find_nodes_with_exits_out_of_container();
            out.node_has_reachable_exit = out
                .node_has_reachable_exit
                .iter()
                .zip(leaving)
                .map(|(a, b)| *a || b)
                .collect();
        }

        out
    }

    pub fn get_node(&self, block: &IrBlock) -> Option<&ControlFlowNode> {
        let index = self.block_index_by_address.get(&block.start_address)?;
        self.nodes.get(*index)
    }

    pub fn has_reachable_exit(&self, node_index: usize) -> bool {
        self.node_has_reachable_exit
            .get(node_index)
            .copied()
            .unwrap_or(false)
    }

    pub fn has_direct_exit_out_of_container(&self, node_index: usize) -> bool {
        self.node_has_direct_exit_out_of_container
            .get(node_index)
            .copied()
            .unwrap_or(false)
    }

    fn initialize_nodes(&mut self, container: &IrBlockContainer) {
        for (index, block) in container.blocks.iter().enumerate() {
            self.nodes
                .push(ControlFlowNode::new(index, Some(block.start_address)));
            self.block_index_by_address
                .insert(block.start_address, index);
        }

        self.node_has_direct_exit_out_of_container = vec![false; self.nodes.len()];
        self.node_has_reachable_exit = vec![false; self.nodes.len()];
    }

    fn create_edges(&mut self, container: &IrBlockContainer) {
        let nested_descendant_blocks = collect_nested_descendant_blocks(container);

        for (index, block) in container.blocks.iter().enumerate() {
            let descendant_container_ids = collect_descendant_container_ids(block);
            for statement in &block.statements {
                self.process_statement(
                    statement,
                    index,
                    &nested_descendant_blocks,
                    &descendant_container_ids,
                );
            }
            self.add_fallthrough_edge_if_needed(block, index);
        }
    }

    fn add_fallthrough_edge_if_needed(&mut self, block: &IrBlock, block_index: usize) {
        if block_has_unreachable_endpoint(block) {
            return;
        }

        let next_index = block_index + 1;
        if next_index < self.nodes.len() {
            self.connect_nodes(block_index, next_index);
            return;
        }

        self.node_has_direct_exit_out_of_container[block_index] = true;
    }

    fn process_statement(
        &mut self,
        statement: &IrStatement,
        source_node_index: usize,
        nested_descendant_blocks: &HashSet<u32>,
        descendant_container_ids: &HashSet<u32>,
    ) {
        match statement {
            IrStatement::Jump(IrJump { target_address }) => {
                self.process_branch_target(
                    *target_address,
                    source_node_index,
                    nested_descendant_blocks,
                );
            }
            IrStatement::If(IrIf {
                true_statement,
                false_statement,
                ..
            }) => {
                self.process_statement(
                    true_statement,
                    source_node_index,
                    nested_descendant_blocks,
                    descendant_container_ids,
                );
                if let Some(false_stmt) = false_statement.as_ref() {
                    self.process_statement(
                        false_stmt,
                        source_node_index,
                        nested_descendant_blocks,
                        descendant_container_ids,
                    );
                }
            }
            IrStatement::Switch(IrSwitch {
                cases,
                default_target,
                ..
            }) => {
                for target in cases.values() {
                    self.process_branch_target(
                        *target,
                        source_node_index,
                        nested_descendant_blocks,
                    );
                }
                if let Some(target) = default_target {
                    self.process_branch_target(
                        *target,
                        source_node_index,
                        nested_descendant_blocks,
                    );
                }
            }
            IrStatement::Leave(IrLeave {
                target_container_id,
            }) => {
                if self.is_leave_out_of_container(*target_container_id, descendant_container_ids) {
                    self.node_has_direct_exit_out_of_container[source_node_index] = true;
                }
            }
            IrStatement::BlockContainer(container) => {
                for nested in &container.blocks {
                    for nested_statement in &nested.statements {
                        self.process_statement(
                            nested_statement,
                            source_node_index,
                            nested_descendant_blocks,
                            descendant_container_ids,
                        );
                    }
                }
            }
            IrStatement::Block(block) => {
                for nested_statement in &block.statements {
                    self.process_statement(
                        nested_statement,
                        source_node_index,
                        nested_descendant_blocks,
                        descendant_container_ids,
                    );
                }
            }
            _ => {}
        }
    }

    fn process_branch_target(
        &mut self,
        target_address: u32,
        source_node_index: usize,
        nested_descendant_blocks: &HashSet<u32>,
    ) {
        if let Some(target_index) = self.block_index_by_address.get(&target_address).copied() {
            self.connect_nodes(source_node_index, target_index);
            return;
        }

        if nested_descendant_blocks.contains(&target_address) {
            return;
        }

        self.node_has_direct_exit_out_of_container[source_node_index] = true;
    }

    fn is_leave_out_of_container(
        &self,
        leave_target_container_id: u32,
        descendant_container_ids: &HashSet<u32>,
    ) -> bool {
        if leave_target_container_id == self.function_body_container_id {
            return false;
        }
        if descendant_container_ids.contains(&leave_target_container_id) {
            return false;
        }
        true
    }

    fn connect_nodes(&mut self, source_index: usize, target_index: usize) {
        if !self.nodes[source_index].successors.contains(&target_index) {
            self.nodes[source_index].successors.push(target_index);
        }
        if !self.nodes[target_index]
            .predecessors
            .contains(&source_index)
        {
            self.nodes[target_index].predecessors.push(source_index);
        }
    }

    fn find_nodes_with_exits_out_of_container(&self) -> Vec<bool> {
        let mut leaving = vec![false; self.nodes.len()];
        for node in &self.nodes {
            if leaving[node.user_index]
                || !self.node_has_direct_exit_out_of_container[node.user_index]
            {
                continue;
            }

            let mut current = Some(node.user_index);
            while let Some(index) = current {
                if leaving[index] {
                    break;
                }
                leaving[index] = true;
                current = self.nodes[index].immediate_dominator;
            }
        }
        leaving
    }
}

fn block_has_unreachable_endpoint(block: &IrBlock) -> bool {
    let Some(last) = block.statements.last() else {
        return false;
    };
    statement_has_unreachable_endpoint(last)
}

fn statement_has_unreachable_endpoint(statement: &IrStatement) -> bool {
    match statement {
        IrStatement::Jump(_) | IrStatement::Leave(_) | IrStatement::Return(IrReturn) => true,
        IrStatement::Switch(IrSwitch { .. }) => true,
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            let Some(false_stmt) = false_statement else {
                return false;
            };
            statement_has_unreachable_endpoint(true_statement)
                && statement_has_unreachable_endpoint(false_stmt)
        }
        IrStatement::Block(block) => block
            .statements
            .last()
            .is_some_and(statement_has_unreachable_endpoint),
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { .. }) => true,
        IrStatement::HighLevelWhile(IrHighLevelWhile { .. }) => false,
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { .. }) => false,
        _ => false,
    }
}

fn collect_nested_descendant_blocks(container: &IrBlockContainer) -> HashSet<u32> {
    let mut descendants = HashSet::<u32>::new();
    let mut visited_containers = HashSet::<u32>::new();

    fn visit_statement(
        statement: &IrStatement,
        descendants: &mut HashSet<u32>,
        visited_containers: &mut HashSet<u32>,
    ) {
        match statement {
            IrStatement::BlockContainer(nested) => {
                visit_container(nested, descendants, visited_containers)
            }
            IrStatement::If(IrIf {
                true_statement,
                false_statement,
                ..
            }) => {
                visit_statement(true_statement, descendants, visited_containers);
                if let Some(false_stmt) = false_statement.as_ref() {
                    visit_statement(false_stmt, descendants, visited_containers);
                }
            }
            IrStatement::Block(block) => {
                for child in &block.statements {
                    visit_statement(child, descendants, visited_containers);
                }
            }
            _ => {}
        }
    }

    fn visit_container(
        nested: &IrBlockContainer,
        descendants: &mut HashSet<u32>,
        visited_containers: &mut HashSet<u32>,
    ) {
        if !visited_containers.insert(nested.id) {
            return;
        }
        for nested_block in &nested.blocks {
            descendants.insert(nested_block.start_address);
            for statement in &nested_block.statements {
                visit_statement(statement, descendants, visited_containers);
            }
        }
    }

    for block in &container.blocks {
        for statement in &block.statements {
            visit_statement(statement, &mut descendants, &mut visited_containers);
        }
    }

    descendants
}

fn collect_descendant_container_ids(block: &IrBlock) -> HashSet<u32> {
    let mut result = HashSet::<u32>::new();
    let mut visited_containers = HashSet::<u32>::new();

    fn visit_statement(
        statement: &IrStatement,
        result: &mut HashSet<u32>,
        visited_containers: &mut HashSet<u32>,
    ) {
        match statement {
            IrStatement::BlockContainer(nested) => {
                if !visited_containers.insert(nested.id) {
                    return;
                }
                result.insert(nested.id);
                for nested_block in &nested.blocks {
                    for nested_statement in &nested_block.statements {
                        visit_statement(nested_statement, result, visited_containers);
                    }
                }
            }
            IrStatement::If(IrIf {
                true_statement,
                false_statement,
                ..
            }) => {
                visit_statement(true_statement, result, visited_containers);
                if let Some(false_stmt) = false_statement.as_ref() {
                    visit_statement(false_stmt, result, visited_containers);
                }
            }
            IrStatement::Block(inner_block) => {
                for inner in &inner_block.statements {
                    visit_statement(inner, result, visited_containers);
                }
            }
            _ => {}
        }
    }

    for statement in &block.statements {
        visit_statement(statement, &mut result, &mut visited_containers);
    }

    result
}
